exec/java-exec/src/main/java/org/apache/drill/exec/metastore/analyze/AnalyzeInfoProviderRegistry.java

```diff
@@ -0,0 +1,128 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.exec.metastore.analyze;
+
+import org.apache.drill.common.scanner.persistence.AnnotatedClassDescriptor;
+import org.apache.drill.common.scanner.persistence.ScanResult;
+import org.apache.drill.exec.physical.base.GroupScan;
+import org.apache.drill.metastore.exceptions.MetastoreException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Class is responsible for dynamic initialization of all implementations
+ * of {@link AnalyzeInfoProvider} annotated with {@link AnalyzeInfoProviderTemplate} and
+ * returning required {@link AnalyzeInfoProvider} instance for specific {@link GroupScan} implementation.
+ * <p>
+ *
+ * AnalyzeInfoProvider list initialization is delayed until {@link #get(GroupScan)} method is called.
+ * AnalyzeInfoProvider implementations must have constructor without arguments.
+ */
+public class AnalyzeInfoProviderRegistry {
+  private static final Logger logger = LoggerFactory.getLogger(AnalyzeInfoProviderRegistry.class);
+
+  private final ScanResult classpathScan;
+  private volatile List<AnalyzeInfoProvider> analyzeInfoProviders;
+
+  public AnalyzeInfoProviderRegistry(ScanResult classpathScan) {
+    this.classpathScan = classpathScan;
+  }
+
+  /**
+   * Returns {@link AnalyzeInfoProvider} instance suitable for specified {@link GroupScan}.
+   *
+   * @param groupScan group scan
+   * @return {@link AnalyzeInfoProvider} instance
+   */
+  public AnalyzeInfoProvider get(GroupScan groupScan) {
+    if (analyzeInfoProviders == null) {
+      synchronized (this) {
+        if (analyzeInfoProviders == null) {
+          analyzeInfoProviders = Collections.unmodifiableList(init());
+        }
+      }
+    }
+
+    List<AnalyzeInfoProvider> suitableProviders = analyzeInfoProviders.stream()
+        .filter(analyzeInfoProvider -> analyzeInfoProvider.supportsGroupScan(groupScan))
+        .collect(Collectors.toList());
+
+    if (suitableProviders.isEmpty()) {
+      throw new MetastoreException(String.format("Unable to find AnalyzeInfoProvider implementation for GroupScan class [%s]",
```

Comment:
Metastore exception is only for Metastore operations and should not be thrown outside Metastore module...